### ISSUE TITLE
Connection timeout

### ISSUE THREAD

**Issue by mna** [IBM-Swift/BlueSocket#65]
Hello,

It looks like there's no way to set a connection timeout right now? There's a read/write timeout that was added (awesome!) but it would be nice to have a connection timeout too.  From what I've read here (http://stackoverflow.com/questions/2597608/c-socket-connection-timeout, second answer is more detailed than accepted one), it involves setting the socket in non-blocking mode and using select with a timeout until the socket is connected (the socket can then be put back in blocking mode if that was how it was requested initially).

I believe the BlueSocket package exposes everything needed to implement this, but this is a non-trivial piece of code and it would make for a much nicer API to have e.g. a `timeout: TimeInterval` (or `UInt` in milliseconds, as is the case for read/write timeouts) parameter added to the `connect` variants.

Thoughts? And thanks for the great package!

Martin

**Comment by SelvaBalasubramanian** [comment]
@mna have u tried this...can you help me out with is case... i need fix a timeout for connect.
https://github.com/IBM-Swift/BlueSocket/issues/76
can you pls check the above issue and help me out....

**Comment by mna** [comment]
Yeah I tried it out and commented on #76, so it looks like a connection timeout cannot be implemented with the current API (without dropping to lower-level socket calls).

**Comment by billabt** [comment]
There's a bit more to it than that.  I'm looking at it...

**Comment by billabt** [comment]
I think I've found the problem.  Hope to have a fix posted over the weekend.  I've got the code complete but haven't finished testing.  Will advise...  With this fix, the new _connect_ API will look like this:
```
public func connect(to host: String, port: Int32, timeout: UInt = 0) throws
```
The `timeout` parameter is ignored if blocking is not turned off for the _Socket_.

**Comment by billabt** [comment]
Decided to enhance the above a bit more.  You won't have to set the Socket into non-blocking mode to use the _timeout_.  You just have to supply a timeout value greater than zero.  Here's the new documentation for the _connect_ API from the README.md:

- `connect(to host: String, port: Int32, timeout: UInt = 0)` - This API allows you to connect to a server based on the `hostname` and `port` you provide. Note: an `exception` will be thrown by this function if the value of `port` is not in the range `1-65535`.  Optionally, you can set `timeout` to the number of milliseconds to wait for the connect. Note: If the socket is in blocking mode it will be changed to non-blocking mode *temporarily* if a `timeout` greater than zero (0) is provided. The returned socket will be *set back to its original setting (blocking or non-blocking)*.

This new API is available in _version 0.12.60_.

**Comment by billabt** [comment]
Going to close this for now.  Feel free to re-open if you run into any issues.  Thanks.
